### PR TITLE
Adding anomaly filtration logic for alerts

### PR DESCRIPTION
**Pull request by puneetjaiswal**
- Added Module to support various type of Filtration Rules 
- TODO: configure filtration rule per metric configuration level

### CODE REVIEW

**Comment by kishoreg** [comment]
Lgtm

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/alert/AnomalyFiltrationHelper.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+      this.threshold = threshold;
+    }
+
+    @Override
+    public boolean isQualified(MergedAnomalyResultDTO anomaly) {
+      Double lengthInHour =
+          (anomaly.getEndTime().doubleValue() - anomaly.getStartTime().doubleValue()) / 36_00_000;
+      Double qualificationScore =
+          Math.pow(lengthInHour, alpha) * Math.pow(Math.abs(anomaly.getWeight()), beta);
+      if (qualificationScore > threshold) {
```
Comment (bl44) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/alert/AnomalyFiltrationHelper.java]: Can use return (qualificationScore > threshold), to replace line 33 to 37, to simply the code.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/datalayer/pojo/MergedAnomalyResultBean.java [hunk 2]
```diff
@@ -107,7 +107,10 @@ public void setCollection(String collection) {
     this.collection = collection;
   }
 
-
+  /**
```
Comment (bl44) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/datalayer/pojo/MergedAnomalyResultBean.java]: Here, you see severity, later you do return weight. It's still confusing. Can you add more comments here for better clarification.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/alert/EmailHelper.java [hunk 3]
```diff
@@ -121,4 +127,39 @@ public static DateTime calculateGraphDataStart(DateTime start, DateTime end,
     }
     return start;
   }
+
+  public static void sendEmailWithTextBody(HtmlEmail email, SmtpConfiguration smtpConfigutation, String subject,
```
Comment (bl44) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/alert/EmailHelper.java]: This function looks the same as the sentEmailWithHtml defined in line 137. Can you consolidate these two?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/alert/AnomalyFiltrationHelper.java (hunk 1 above)
Comment (puneetjaiswal) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/alert/AnomalyFiltrationHelper.java]: sure.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/datalayer/pojo/MergedAnomalyResultBean.java (hunk 2 above)
Comment (puneetjaiswal) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/datalayer/pojo/MergedAnomalyResultBean.java]: in thirdeye weight = severity. I will add more comments.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/alert/EmailHelper.java (hunk 3 above)
Comment (puneetjaiswal) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/alert/EmailHelper.java]: sure. It was result of merging two util classes into one.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/alert/EmailHelper.java [hunk 4]
```diff
@@ -121,4 +127,39 @@ public static DateTime calculateGraphDataStart(DateTime start, DateTime end,
...
+        email.setSSLOnConnect(true);
+      }
+      email.setFrom(fromAddress);
+      for (String address : toAddress.split(EMAIL_ADDRESS_SEPARATOR)) {
+        email.addTo(address);
+      }
+      email.send();
+      LOG.info("Sent!");
+    } else {
+      LOG.error("No failure email configs provided!");
```
Comment (bl44) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/alert/EmailHelper.java]: "No failure email configs provided" seems too generic. For example 'alert email configuration'?, just try to separate it from other 'failure email configs' log, if there is any.

Comment (puneetjaiswal) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/alert/EmailHelper.java]: sure.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/alert/AlertTaskRunner.java [hunk 5]
```diff
@@ -139,7 +145,24 @@ private void runTask() throws Exception {
...
-    updateNotifiedStatus(results);
+    updateNotifiedStatus(allResults);
+  }
+
+  private List<MergedAnomalyResultDTO> applyFiltrationRule(List<MergedAnomalyResultDTO> results) {
+    if (results.size() == 0) {
+      return results;
+    }
+    // TODO: configure filtration rule at metric level and fetch it from DB
+    AnomalyFiltrationHelper.FiltrationRule filtrationRule =
```
Comment (bl44) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/alert/AlertTaskRunner.java]: For a given MergedAnomalyResultDTO, multiple rules can be applied. Want to make sure that  you can get a list of filtration Rules, based on metric_id and/or pattern type.  For simplicity let's do metric first, but we can have at least more than 1 rules.

Comment (puneetjaiswal) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/alert/AlertTaskRunner.java]: Sure, I will update TODO comment.

**Pull request merged**